keps/sig-cli/20190805-pod-troubleshooting.md

```diff
@@ -0,0 +1,380 @@
+implemented by the Windows container runtime, and Windows containers cannot be
+built `FROM scratch` so debugging tools are already available in the container.
+While not excluding Windows containers, this KEP will focus on the Linux use
+case.
+
+[distroless container image]: https://github.com/GoogleCloudPlatform/distroless
+[distroless for k8s system images]: https://git.k8s.io/enhancements/keps/sig-release/20190316-rebase-images-to-distroless.md
+[scratch debugger]: https://github.com/kubernetes-retired/contrib/tree/master/scratch-debugger
+
+## Proposal
+
+We will add a new command `kubectl debug` that will:
+
+1. Construct a `v1.Container` for the debug container based on command line
+   arguments. This will include an optional container for namespace targeting
+   (described below).
+2. Fetch the specified pod's existing ephemeral containers using
+   `GetEphemeralContainers()` in the generated pod client.
+3. Append the new debug container to the pod's ephemeral containers and call
+   `UpdateEphemeralContainers()`.
+4. [optional] Watch pod for updates to the debug container's `ContainerStatus`
+   and automatically attach once the container is running.
+
+### Debug Container Naming
+
+Currently, there is no support for deleting or recreating ephemeral containers.
+In cases where the user does not specify a name, `kubectl` will generate a
+unique name and display it to the user.
+
+### Container Namespace Targeting
+
+In order to see processes running in other containers in the pod, [Process
+Namespace Sharing] should be enabled for the pod. In cases where process
+namespace sharing isn't enabled for the pod, `kubectl` will set
+`TargetContainer` in the `EphemeralContainer`. This will cause the ephemeral
+container to be created in the namespaces of the target container in runtimes
+that support container namespace targeting.
+
+[Process Namespace Sharing]: https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace
+
+### Interactive Troubleshooting and Automatic Attaching
+
+Since the primary use case for `kubectl debug` is interactive troubleshooting,
+`kubectl debug` will automatically attach to the console of the newly created
+ephemeral container and will default to creating containers with `Stdin` and
+`TTY` enabled.
+
+These may be disabled via command line flags.
+
+### Proposed `kubectl debug` arguments
+
+```
+% kubectl help debug
+Execute a container in a pod.
+
+Examples:
+  # Get network address configuration from pod mypod
+  kubectl debug mypod ip addr list
+
+  # Start an interactive debugging session with a debian image
+  kubectl debug mypod --image=debian
+
+Options:
+  -a, --attach=true: Automatically attach to container once created.
+  -c, --container='': Container name. If omitted one will be chosen.
+  -i, --stdin=true: Pass stdin to the container
+  -m, --image='busybox': Container image to use for debug container.
```

Comment:
`-m` might be confusing, I'd drop short for now. Also I'm inclined not to default this to anything, just require a user to always pass this.

Reply:
SG, done.